pycqed/measurement/multi_qubit_module.py

```diff
@@ -2017,45 +2027,64 @@ def calibrate_n_qubits(qubits, f_LO, sweep_points_dict, sweep_params=None,
                                     'updated.' % e)
 
 
-def measure_chevron(qbc, qbt, qbr, hard_sweep_params, soft_sweep_params,
-                    cz_pulse_name, upload=True, label=None,
+def measure_chevron(dev, qbc, qbt, hard_sweep_params, soft_sweep_params,
+                    cz_pulse_name, upload=True, label=None, qbr=None,
                     classified=False, n_cal_points_per_state=2,
-                    cal_states='auto', prep_params=None,
-                    exp_metadata=None, analyze=True):
+                    num_cz_gates=1, cal_states='auto', prep_params=None,
+                    exp_metadata=None, analyze=True, return_seq=False):
+
+    if isinstance(qbc, str):
+        qbc = dev.get_qb(qbc)
+    if isinstance(qbt, str):
+        qbt = dev.get_qb(qbt)
+
+    if qbr is None:
+        qbr = qbt
+    elif qbr != qbc and qbr != qbt:
+        raise ValueError('Only target or control qubit can be read out!')
+
+    # check whether qubits are connected
+    dev.check_connection(qbc, qbt)
 
     if len(list(soft_sweep_params)) > 1:
         log.warning('There is more than one soft sweep parameter.')
     if label is None:
         label = 'Chevron_{}{}'.format(qbc.name, qbt.name)
-    MC = qbr.instr_mc.get_instr()
-    for qb in [qbc, qbt, qbr]:
+    MC = dev.find_instrument('MC')
+    for qb in [qbc, qbt]:
         qb.prepare(drive='timedomain')
 
     cal_states = CalibrationPoints.guess_cal_states(cal_states)
     cp = CalibrationPoints.single_qubit(qbr.name, cal_states,
                                         n_per_state=n_cal_points_per_state)
 
     if prep_params is None:
-        prep_params = \
-            get_multi_qubit_prep_params([qb.preparation_params()
-                                         for qb in [qbc, qbt]])
+        prep_params = dev.get_prep_params([qbc, qbt])
+
+    operation_dict = dev.get_operation_dict()
+
     sequences, hard_sweep_points, soft_sweep_points = \
         fsqs.chevron_seqs(
             qbc_name=qbc.name, qbt_name=qbt.name, qbr_name=qbr.name,
             hard_sweep_dict=hard_sweep_params,
             soft_sweep_dict=soft_sweep_params,
-            operation_dict=get_operation_dict([qbc, qbt, qbr]),
+            operation_dict=operation_dict,
             cz_pulse_name=cz_pulse_name,
+            num_cz_gates=num_cz_gates,
             cal_points=cp, upload=False, prep_params=prep_params)
 
+    if return_seq:
+        return sequences
+
     hard_sweep_func = awg_swf.SegmentHardSweep(
         sequence=sequences[0], upload=upload,
         parameter_name=list(hard_sweep_params)[0],
         unit=list(hard_sweep_params.values())[0]['unit'])
     MC.set_sweep_function(hard_sweep_func)
     MC.set_sweep_points(hard_sweep_points)
 
-    channels_to_upload = [qbc.get_operation_dict()[cz_pulse_name]['channel']]
+    # sweep over flux pulse amplitude of qbc
+    channels_to_upload = [qbc.flux_pulse_channel()]
```

Comment:
Always correct to upload only one channel? What if cz_pulse_name is set to a FLIP gate? Couldn't the user then decide to also put parameters of the other fluxed qubits into the sweep_points?

Reply:
Agree! Add new channels_to_upload parameter, that contains the names of the channels that should be reuploaded during the soft sweep and which is set to the flux pulse channels of qbt and qbc by default.